### PR TITLE
Add CI to validate vulnerability format

### PR DESCRIPTION
**Pull request by vdeturckheim**
This PR aims at adding a CI step to enforce a format for vulnerability. I also fixed a few issues in vuln files.

### CODE REVIEW

File: tools/vuln_valid/index.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+'use strict';
+const Joi = require('joi');
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Can you remove the hapi'isms from the code style (capitalized imports, four space indentation, empty line at function start, etc.).

File: tools/vuln_valid/index.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+const Path = require('path');
+const Fs = require('fs');
+
+const vulnPath = Path.join(__dirname, '..', '..', 'vuln');
+const coreVulnPath = Path.join(vulnPath, 'core');
+const npmVulnPath = Path.join(vulnPath, 'npm');
+
+const coreModel = Joi.object().keys({
+    cve: Joi.array().items(Joi.string().regex(/CVE-\d{4}-\d+/)),
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Can you add explicit `required()` and `optional()` calls to the schema. Right now you have some `optional()`s, but I believe that is the default anyway, so the intentions are not really clear.

File: tools/vuln_valid/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    references: Joi.string().allow('').allow(null),
+    cvss_vector: Joi.string().allow('').allow(null),
+    cvss_score: Joi.number().allow(null),
+    coordinating_vendor: Joi.string().allow('')
+});
+
+let fail = false;
+
+Fs.readdirSync(coreVulnPath)
+    .map((x) => Path.join(coreVulnPath, x))
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Can you use a more meaningful variable name than `x`.

Comment (cjihrig) [reply tools/vuln_valid/index.js]: Actually, there's no need for the `map()`. You aren't using the result, so just perform the `Path.join()` in the `forEach()`.

File: tools/vuln_valid/index.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    .forEach((path) => {
+
+        const vuln = JSON.parse(Fs.readFileSync(path));
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Is the intention to process all entries, even on error? If so, this needs a `try...catch`.

File: tools/vuln_valid/index.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        const result = Joi.validate(vuln, coreModel);
+        if (result.error) {
+            console.log(`File ${path}:`);
+            console.log(result.error);
+            fail = true;
+        }
+    });
+
+Fs.readdirSync(npmVulnPath)
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Same comments apply here.

File: tools/vuln_valid/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+Fs.readdirSync(coreVulnPath)
+    .map((x) => Path.join(coreVulnPath, x))
+    .forEach((path) => {
+
+        const vuln = JSON.parse(Fs.readFileSync(path));
+        const result = Joi.validate(vuln, coreModel);
+        if (result.error) {
+            console.log(`File ${path}:`);
+            console.log(result.error);
+            fail = true;
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Instead of using a `fail` variable, you can just set `process.exitCode = 1;`

File: tools/vuln_valid/index.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+const Fs = require('fs');
+
+const vulnPath = Path.join(__dirname, '..', '..', 'vuln');
+const coreVulnPath = Path.join(vulnPath, 'core');
+const npmVulnPath = Path.join(vulnPath, 'npm');
+
+const coreModel = Joi.object().keys({
+    cve: Joi.array().items(Joi.string().regex(/CVE-\d{4}-\d+/)),
+    ref: Joi.string().uri().optional(),
+    vulnerable: Joi.string(),
```
Comment (cjihrig) [inline tools/vuln_valid/index.js]: Would it be possible to create a joi extension that utilizes the `semver` module to validate these? If so, that would be cool.

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+node_modules/
```
Comment (mcollina) [inline .gitignore]: We should adopt a larger `.gitignore`, e.g. https://github.com/github/gitignore/blob/master/Node.gitignore

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+language: node_js
+
+node_js:
+  - "node"
```
Comment (mcollina) [inline .travis.yml]: we should pin a version here

**Comment by mhdawson** [comment]
There are no platform concerns so using Travis is probably ok, but just wondering if there is any reason we cannot just use a job in the Node.js CI ?

**Comment by vdeturckheim** [comment]
@mhdawson no real reason indeed. Mostly because enabling travis feels more straightforward to me.

**Comment by lirantal** [comment]
If it's taking some time to use the Node.js CI maybe we can just start off with travis and then move? it isn't a big deal, and will help us lint all the upcoming vulnerability reports.

On a similar note, if you add this small utility, and another one to convert H1 to this format, and then another one to request a CVE as reed recently pointed out, I thought to open a public npm organization and new repo(s) to organize some of this tooling?

**Comment by mhdawson** [comment]
@lirantal agreed if we can get things up an running quickly with travis and them move later if appropriate.

**Comment by mhdawson** [comment]
In terms of npm, we already have a number of modules and what we have done is to keep them as repos under the nodejs org.  That would probably be my first suggestion.  Generally, they start out in an individuals repo and then we move them over once they have developed to the point where they have regular use in the project.  On the npm side we'd don't have an organization but do have a nodejs user that we ask to be added as a collaborator to all modules that come from repos under the nodejs org.  Publishing is still done by individuals but that user allows us to recover if all of the collaborators drop out of the community for some reason. See: https://github.com/nodejs/build/blob/master/doc/process/npm_management.md

**Comment by vdeturckheim** [comment]
Awesome! I will merge + enable travis tomorrow.

**Comment by lirantal** [comment]
@mhdawson oh thanks for that info, I will consolidate some of the tooling into a new issue and we can further discuss.

**Comment by vdeturckheim** [comment]
@mhdawson seems that I can't add Travis to the org :/ we'll probably have to use Jenkins if we want to do this.

**Comment by mhdawson** [comment]
@vdeturckheim I can create a blank Jenkins job for you to configure.  If there is a script in our repo that we just need to run I could also just create the job and point it to that script.  Let me know which works for you.

**Comment by vdeturckheim** [comment]
@mhdawson I think the best solution is to have a script in this repo. Is it possible to have a CI job that just runs `$ npm test` ?

**Comment by mhdawson** [comment]
For something that is checked out of this repo, yes.  We don't want to run just anything but as long as we know what the module is (and ideally it is under our control) it will not be a problem.

**Comment by gibfahn** [comment]
>@mhdawson I think the best solution is to have a script in this repo. Is it possible to have a CI job that just runs $ npm test ?

I'd suggest just adding a Jenkinsfile (equivalent of a `travis.yml`) to this repo and using that. We can configure the job to just run the pipeline.

@vdeturckheim are you at Index today and tomorrow? If so we could take some time to go through it and try to get it set up.

**Comment by vdeturckheim** [comment]
@gibfahn Good idea! 
Sadly I was only at Index yesterday :/ I'll try to create a Jenkinsfile, is there already an example in the org somewhere?

**Comment by gibfahn** [comment]
@vdeturckheim sadly not any that really match what you're doing here. The ones we have at the moment are in https://github.com/nodejs/build/tree/master/jenkins/pipelines.

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,6 @@
+language: node_js
+
+node_js:
+  - "8.9.0"
```
Comment (ChALkeR) [inline .travis.yml]: Why not just `"8"`?

**Comment by ChALkeR** [comment]
@vdeturckheim As most (if not all) concerns hare are about the specific CI setup path, perhaps you could split this PR in two — one just introducing `npm test` (and fixes), and the other actually setting up a CI (when we would figure out what that could be)?

Even being available to run `npm test` prior to filing a PR would be a great improvement here.

**Comment by lirantal** [comment]
Agree with @ChALkeR.
Another thing I want to bring up is that maybe the tooling can be managed in npm as its own package? For a quick start we can merge but I see other tooling that we'll build around the whole process which will benefit from several shared capabilities like the schema validation.

**Comment by dgonzalez** [comment]
Agreed.

**Comment by ChALkeR** [comment]
@vdeturckheim 390 from #162 fails the lint.

**Pull request merged**